### DISCUSSION TITLE
Why is zod validation error coming in a JSON string instead of an Array in link error callback?

### DISCUSSION THREAD

**Discussion by capaj** [trpc/trpc#3068] [Q&A]
I have a custom validation using `refine`
```ts
 .input(
      z.object({
        user: z.object({
          email: z
            .string()
            .email()
            .refine(
              (email) =>
                emailProviders.find((tld: string) => {
                  return email.endsWith(tld)
                }) === undefined,
              {
                message: t`Domain blocked, you must use a work email instead of personal one`
              }
            ),
          password: passwordZodType,
          first_name: z.string(),
          last_name: z.string()
        }),
```

and on FE I have a custom link like this:
![image](https://user-images.githubusercontent.com/1305378/197511981-251c9088-c056-4777-97ea-0d5420084970.png)

Why does the log output this?
```
message [
  {
    "code": "custom",
    "message": "Domain blocked, you must use a work email instead of personal one",
    "path": [
      "user",
      "email"
    ]
  }
] string
```

couldn't it come as an array into the link `error` already? Feels like trying to parse every error that might get passed from BE is not optimal in this place. Maybe TRPC knows when a validation fails(400 http status code?) and in those cases it could parse it before link callback is invoked?

**Comment by KATT** [comment]
We're just passing whatever Zod has as an `error.message`, if you want more attributes propagated from zod you can use custom error formatting: https://trpc.io/docs/v10/error-formatting
